File: python/src/0145.py
# [145. Binary Tree Postorder Traversal (Easy)](https://leetcode.com/problems/binary-tree-postorder-traversal/)
# @topic Stack
# @topic Tree
# @topic Depth-First Search
# @topic Binary Tree
#
# Constraints:
# The number of the nodes in the tree is in the range [0, 100].
# -100 <= Node.val <= 100

from collections import deque
from typing import Optional

from structs.BinaryTree import create, TreeNode


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    # def postorderTraversal(self, root: Optional[TreeNode]) -> list[int]:
    #     def dfs(root, res):
    #         if root:
    #             root.left and dfs(root.left, res)
    #             root.right and dfs(root.right, res)
    #             res.append(root.val)
    #         return res
    #     return dfs(root, []) if root else []

    def postorderTraversal(self, root: Optional[TreeNode]) -> list[int]:
        res = []
        pre = None
        cur = root
        stk = deque()
        while cur or stk:
            while cur:
                stk.append(cur)
                cur = cur.left
            cur = stk[-1]
            if cur.right and cur.right != pre:
                cur = cur.right
            else:
                stk.pop()
                res.append(cur.val)
                pre = cur
                cur = None
        return res

    # def postorderTraversal(self, root: Optional[TreeNode]) -> list[int]:
    #     res = []
    #     cur = root
    #     while cur:
    #         if pre := cur.right:
    #             while pre.left and pre.left != cur:
    #                 pre = pre.left
    #             if not pre.left:
    #                 res.append(cur.val)
    #                 pre.left = cur
    #                 cur = cur.right
    #             else:
    #                 pre.left = None
    #                 cur = cur.left
    #         else:
    #             res.append(cur.val)
    #             cur = cur.left
    #     return res[::-1]

    # def postorderTraversal(self, root: Optional[TreeNode]) -> list[int]:
    #     res = []
    #     cur = root
    #     while cur:
    #         res.append(cur.val)
    #         if pre := cur.right:
    #             while pre.left:
    #                 pre = pre.left
    #             pre.left = cur.left
    #             pre = cur
    #             cur = cur.right
    #             pre.left = None
    #             pre.right = None
    #         else:
    #             cur = cur.left
    #     return res[::-1]


if __name__ == '__main__':
    import unittest
    from timeit import timeit

    CASES = (
        (create(TreeNode, 1, None, 2, 3), [3, 2, 1]),
        (None, []),
        (TreeNode(1), [1]),
    )

    class Test(unittest.TestCase):

        def test(self):
            postorderTraversal = Solution().postorderTraversal
            for root, excepted in CASES:
                actual = postorderTraversal(root)
                self.assertEqual(actual, excepted)

    unittest.main()

    postorderTraversal = Solution().postorderTraversal
    for root, excepted in CASES:
        print(timeit(lambda: postorderTraversal(root), number=100000))

File: cpp/src/0145.cpp
/*
[145. Binary Tree Postorder Traversal (Easy)](https://leetcode.com/problems/binary-tree-postorder-traversal/)

@topic: Stack
@topic: Tree
@topic: Depth-First Search
@topic: Binary Tree

Constraints:
  - The number of the nodes in the tree is in the range `[0, 100]`.
  - `-100 <= Node.val <= 100`
*/


#include "structs/BinaryTree.hpp"

class Solution {
   public:
    vector<int> postorderTraversal(TreeNode *root) {
        vector<int> res;
        TreeNode *pre;
        TreeNode *cur = root;
        while (cur) {
            if ((pre = cur->right)) {
                while (pre->left && pre->left != cur) {
                    pre = pre->left;
                }
                if (pre->left == nullptr) {
                    res.emplace_back(cur->val);
                    pre->left = cur;
                    cur = cur->right;
                } else {
                    pre->left = nullptr;
                    cur = cur->left;
                }
            } else {
                res.emplace_back(cur->val);
                cur = cur->left;
            }
        }
        reverse(res.begin(), res.end());
        return res;
    }

    // vector<int> postorderTraversal(TreeNode *root) {
    //     if (!root) return {};
    //     vector<int> res;
    //     stack<TreeNode *> stk;
    //     TreeNode *pre = nullptr;
    //     TreeNode *cur = root;
    //     while (cur || !stk.empty()) {
    //         while (cur) {
    //             stk.push(cur);
    //             cur = cur->left;
    //         }
    //         cur = stk.top();
    //         if (cur->right && cur->right != pre) {
    //             cur = cur->right;
    //         } else {
    //             stk.pop();
    //             res.emplace_back(cur->val);
    //             pre = cur;
    //             cur = nullptr;
    //         }
    //     }
    //     return res;
    // }
};

#include <cassert>
#include <chrono>
#include <iostream>

using std::tuple;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    using namespace BinaryTree;

    Solution o;

    vector<tuple<TreeNode*, vector<int>>> CASES = {
        {of({1, INT_MIN, 2, 3}), {3, 2, 1}},
        {nullptr, {}},
        {of({1}), {1}},
    };

    for (auto& [root, excepted] : CASES) {
        assert(o.postorderTraversal(root) == excepted);
    }

    auto start = system_clock::now();
    for (auto& [root, _] : CASES) {
        for (auto i = 0; i < 100000; ++i) {
            o.postorderTraversal(root);
        }
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;

}
